### PR TITLE
HDDS-11410. Refactoring more tests from TestContainerBalancerTask

### PR DESCRIPTION
**Pull request by Montura**
In [PR for HDDS-9889](https://github.com/apache/ozone/pull/5758) we [discussed](https://github.com/apache/ozone/pull/5758#pullrequestreview-1963251436) with [Siddhant Sangwan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=siddhant) that tests form `org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerTask` could be refactored using `MockedSCM` class (introduced in [HDDS-9889](https://issues.apache.org/jira/browse/HDDS-9889))

Some work has been already done in:
1. https://github.com/apache/ozone/pull/5758
2. https://github.com/apache/ozone/pull/6537
3. https://github.com/apache/ozone/pull/6734

## What changes were proposed in this pull request?

1. Refactor some tests from `org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerTask` 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-11410

## How was this patch tested?

Use standalone tests

### CODE REVIEW

**Comment by Montura** [comment]
@Tejaskriya, take a look please

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/balancer/TestContainerBalancerDatanodeNodeLimit.java [hunk 1]
```diff
@@ -208,15 +186,7 @@ public void balancerShouldObeyMaxSizeLeavingSourceLimit(@Nonnull MockedSCM mocke
   @ParameterizedTest(name = "MockedSCM #{index}: {0}")
   @MethodSource("createMockedSCMs")
   public void initializeIterationShouldUpdateUnBalancedNodesWhenThresholdChanges(@Nonnull MockedSCM mockedSCM) {
-    ContainerBalancerConfiguration config = new OzoneConfiguration().getObject(ContainerBalancerConfiguration.class);
-    int nodeCount = mockedSCM.getCluster().getNodeCount();
-    if (nodeCount < DATANODE_COUNT_LIMIT_FOR_SMALL_CLUSTER) {
-      config.setMaxDatanodesPercentageToInvolvePerIteration(100);
-    }
-    config.setThreshold(10);
```
Comment (Tejaskriya) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/balancer/TestContainerBalancerDatanodeNodeLimit.java:216]: Setting the threshold has been removed here and in few more tests below, but in the newly introduced ContainerBalancerConfigBuilder, we aren't setting it. Is this intentional?

**Review by Tejaskriya — COMMENTED** [review]
Overall the PR seems good to me, just a small comment below.

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/balancer/TestContainerBalancerDatanodeNodeLimit.java (hunk 1 above)
Comment (Montura) [reply hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/balancer/TestContainerBalancerDatanodeNodeLimit.java:216]: @Tejaskriya , thank for a review. I forgot about **threshold** thanks, I'll fix it

**Comment by Montura** [comment]
@adoroszlai , you mean explicitly write ```config.setThreshold(...)``` in each test case?

**Comment by adoroszlai** [comment]
@Montura no, I only meant that based on [this conversation](https://github.com/apache/ozone/pull/7156#discussion_r1744938637) something needs to be done about setting threshold.  If you have addressed that, please reply `/ready` to remove the `pending` label.

**Comment by Montura** [comment]
> Marking this issue as un-mergeable as requested.
> 
> Please use `/ready` comment when it's resolved.
> 
> Please note that the PR will be closed after 21 days of inactivity from now. (But can be re-opened anytime later...)
> 
> > set threshold in tests

/ready

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/balancer/TestContainerBalancerDatanodeNodeLimit.java (hunk 1 above)
Comment (Montura) [reply hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/balancer/TestContainerBalancerDatanodeNodeLimit.java:216]: /ready

**Review by Tejaskriya — APPROVED** [review]
LGTM, Thanks for the patch @Montura

**Comment by Montura** [comment]
> Looks like this increases test run time from:
> 
> ```
> [INFO] Tests run: 240, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 34.125 s - in org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit
> [INFO] Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 5.686 s - in org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerTask
> ```
> 
> to:
> 
> ```
> [INFO] Tests run: 304, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 70.262 s - in org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit
> [INFO] Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.735 s - in org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerTask
> ```
> 
> So 2x for `TestContainerBalancerDatanodeNodeLimit` and x/2 for `TestContainerBalancerTask`, but the problem is that the increase is an order of magnitude more.
> 
> The following test cases take more than 1 second:
> 
> ```
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultException(MockedSCM)[1]  Time elapsed: 1.028 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultException(MockedSCM)[2]  Time elapsed: 1.024 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultException(MockedSCM)[3]  Time elapsed: 1.024 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultException(MockedSCM)[4]  Time elapsed: 1.025 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultException(MockedSCM)[5]  Time elapsed: 1.025 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultException(MockedSCM)[6]  Time elapsed: 1.025 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultTimeout(MockedSCM)[1]  Time elapsed: 1.026 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultTimeout(MockedSCM)[2]  Time elapsed: 1.024 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultTimeout(MockedSCM)[3]  Time elapsed: 1.029 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultTimeout(MockedSCM)[4]  Time elapsed: 1.026 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultTimeout(MockedSCM)[5]  Time elapsed: 1.029 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultTimeout(MockedSCM)[6]  Time elapsed: 1.03 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultTimeout(MockedSCM)[7]  Time elapsed: 1.035 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultTimeout(MockedSCM)[8]  Time elapsed: 1.05 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultTimeout(MockedSCM)[9]  Time elapsed: 1.039 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultTimeout(MockedSCM)[10]  Time elapsed: 1.078 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultTimeout(MockedSCM)[11]  Time elapsed: 1.046 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultTimeout(MockedSCM)[12]  Time elapsed: 1.053 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultTimeout(MockedSCM)[13]  Time elapsed: 1.062 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultTimeout(MockedSCM)[14]  Time elapsed: 1.074 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultTimeout(MockedSCM)[15]  Time elapsed: 1.094 s
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit.checkIterationResultTimeout(MockedSCM)[16]  Time elapsed: 1.157 s
> ```

There are 3 possible solutions:
 1. to reduce the total amount of arguments for `checkIterationResultTimeout` and `checkIterationResultException` tests. 
 2. to decrease the `timeout` time limit
 3. Do items 1 and 2 simultaneously

**Comment by adoroszlai** [comment]
> There are 3 possible solutions:
> 
> 1. to reduce the total amount of arguments for `checkIterationResultTimeout` and `checkIterationResultException` tests.
> 2. to decrease the `timeout` time limit
> 3. Do items 1 and 2 simultaneously

Can you please explore how much `timeout` can be decreased without introducing flakiness?  Depending on that, we'd need to go with one of the solutions.

**Comment by Montura** [comment]
UPD:
- `initializeIterationShouldUpdateUnBalancedNodesWhenThresholdChanges` : reduce iterations 50 -> 10
- `checkIterationResultTimeout` : increase `maxEnteringSize`, decrease `timeouts`

**Comment by adoroszlai** [comment]
Thanks @Montura for reducing test time:

```
Tests run: 304, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 41.946 s - in org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.542 s - in org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerTask
```

**Comment by Montura** [comment]
I'll take a look at failed `checkIterationResultException` test

**Comment by adoroszlai** [comment]
> https://github.com/unisonteam/ozone/actions/runs/12050928627
> TestContainerBalancerDatanodeNodeLimit#checkIterationResultException[master]-10x10

Thanks.  `Git Ref (Branch/Commit_ID/Tag)` should be set to `container-balancer-tests`, but here it was testing `master`, where `TestContainerBalancerDatanodeNodeLimit` does not exist.

Normally the workflow would report "no tests were run" error in such case.  However, `master` in your fork (11850f5f9ba6d21f1665b2c07a1ebb73a8623cfa) is older than the commit where this sanity check was introduced (4195ef57a3a29aab3a79747147d0d73dc63c34af).

Can you please launch a new run, with `Git Ref (Branch/Commit_ID/Tag): container-balancer-tests`?

**Comment by Montura** [comment]
Sure, I'll restart

**Comment by adoroszlai** [comment]
@Montura, thank you.  Much better:

```
[INFO] Tests run: 16, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 5.162 s - in org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit
```

https://github.com/unisonteam/ozone/actions/runs/12056169848/job/33618737808#step:7:1654

**Comment by Montura** [comment]
Ready to merge?

**Comment by adoroszlai** [comment]
@siddhantsangwan @sumitagrawl would you like to take a look?

**Comment by adoroszlai** [comment]
Thanks @Montura for the patch.

> @siddhantsangwan @sumitagrawl would you like to take a look?

I guess not.

**Comment by ivandika3** [comment]
Seems TestContainerBalancerDatanodeNodeLimit is flaky 

https://github.com/apache/ozone/actions/runs/12151251117/job/33885814517

**Comment by adoroszlai** [comment]
Thanks @ivandika3 for the info.  Launched a [run](https://github.com/adoroszlai/ozone/actions/runs/12152868992) for the entire test class.

[Previous run](https://github.com/unisonteam/ozone/actions/runs/12056169848) only covered a single method, there may be interaction between test cases.

**Comment by Montura** [comment]
@ivandika3 , which test (or tests) is (are) flaky now?

I can look/run locally and prepare hotfix or tune the limits (as I've done before)

**Comment by adoroszlai** [comment]
@Montura `TestContainerBalancerDatanodeNodeLimit#checkIterationResultException`, when we run the test class, not just a method.

https://github.com/adoroszlai/ozone/actions/runs/12152868992/job/33890752644#step:3:21

**Comment by Montura** [comment]
> @Montura `TestContainerBalancerDatanodeNodeLimit#checkIterationResultException`, when we run the test class, not just a method.
> 
> https://github.com/adoroszlai/ozone/actions/runs/12152868992/job/33890752644#step:3:21

I'll take a look today

**Comment by adoroszlai** [comment]
Thanks @Montura.  Created HDDS-11855 for the failure.

**Comment by Montura** [comment]
@adoroszlai , I'm still investigating. For now the test still is flacky 
https://github.com/unisonteam/ozone/actions/runs/12250923249/job/34175235709

**Pull request merged**